File: java-to-go-projects/17-kv-store/Persistence.java
package kvstore;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.*;
import java.nio.file.*;
import java.util.*;

/**
 * Handles saving and loading the KV store to/from disk.
 * Uses JSON format for human-readable persistence.
 */
public class Persistence {

    private static final ObjectMapper mapper = new ObjectMapper();
    private final String filePath;

    public Persistence(String filePath) {
        this.filePath = filePath;
    }

    /**
     * Save the store's data to disk as JSON.
     */
    public void save(Store store) throws IOException {
        Map<String, String> data = store.snapshot();
        String json = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(data);
        Files.writeString(Path.of(filePath), json);
        System.out.printf("Saved %d entries to %s%n", data.size(), filePath);
    }

    /**
     * Load data from disk into the store (replaces existing data).
     */
    public void load(Store store) throws IOException {
        if (!Files.exists(Path.of(filePath))) {
            System.out.println("No save file found: " + filePath);
            return;
        }

        String json = Files.readString(Path.of(filePath));
        Map<String, String> data = mapper.readValue(json, new TypeReference<>() {});
        store.replaceAll(data);
        System.out.printf("Loaded %d entries from %s%n", data.size(), filePath);
    }
}

File: java-to-go-projects/17-kv-store/rust/src/persistence.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

pub fn save(data: &HashMap<String, String>, file_path: &str) -> Result<(), String> {
    serde_json::to_string_pretty(data)
        .map_err(|e| e.to_string())
        .and_then(|json| {
            fs::write(file_path, json).map_err(|e| format!("Failed to save data: {}", e))
        })
        .map(|_| println!("Data successfully saved to {}", file_path))
}

pub fn load(file_path: &str) -> Result<HashMap<String, String>, String> {
    let path = Path::new(file_path);
    if !path.exists() {
        return Ok(HashMap::new());
    }
    let data = fs::read_to_string(file_path).map_err(|e| format!("Failed to read data: {}", e))?;
    serde_json::from_str(&data)
        .map_err(|e| format!("Failed to parse data: {}", e))
        .map(|map: HashMap<String, String>| {
            println!("Data successfully loaded from {}", file_path);
            map
        })
}
